Framework/HistogramData/src/Interpolate.cpp

```diff
@@ -0,0 +1,220 @@
+  }
+  const auto ysize = input.y().size();
+  if (stepSize >= ysize) {
+    throw std::runtime_error("interpolate - Step size must be smaller "
+                             "than the number of points");
+  }
+  // First and last points are always assumed to be calculated
+  const size_t ncalc = numberCalculated(ysize, stepSize);
+  if (ncalc < minCalculated) {
+    std::ostringstream os;
+    os << "interpolate - " << method << " requires " << minCalculated
+       << " points but only " << ncalc << " were found.";
+    throw std::runtime_error(os.str());
+  }
+}
+
+/**
+ * Perform linear interpolation assuming the input data set to be a Histogram
+ * with XMode=Points. It is assumed all sanity cheks have been performed
+ * by the interpolateLinear entry point.
+ * @param input See interpolateLinear
+ * @param stepSize See interpolateLinear
+ * @param ynew A reference to the output Y values
+ */
+void interpolateYLinearInplace(const Histogram &input, const size_t stepSize,
+                               HistogramY &ynew) {
+  const auto xold = input.points();
+  const auto &yold = input.y();
+  const auto nypts = yold.size();
+  size_t step(stepSize), index2(0);
+  double x1(0.), x2(0.), y1(0.), y2(0.), overgap(0.);
+  // Copy over end value skipped by loop
+  ynew.back() = yold.back();
+  for (size_t i = 0; i < nypts - 1; ++i) // Last point has been calculated
+  {
+    const double xp = xold[i];
+    if (step == stepSize) {
+      x1 = xp;
+      index2 = ((i + stepSize) >= nypts ? nypts - 1 : (i + stepSize));
+      x2 = xold[index2];
+      overgap = 1.0 / (x2 - x1);
+      y1 = yold[i];
+      y2 = yold[index2];
+      step = 1;
+      ynew[i] = yold[i];
+      continue;
+    }
+    // Linear interpolation
+    ynew[i] = (xp - x1) * y2 + (x2 - xp) * y1;
+    ynew[i] *= overgap;
+    step++;
+  }
+}
+
+/**
+ * Perform cubic spline interpolation. It is assumed all sanity checks have been
+ * performed by the interpolateCSpline entry point.
+ * @param input See interpolateCSpline
+ * @param stepSize See interpolateCSpline
+ * @param ynew A reference to the output Y values
+ */
+void interpolateYCSplineInplace(const Histogram &input, const size_t stepSize,
+                                HistogramY &ynew) {
+  const auto xold = input.points();
+  const auto &yold = input.y();
+  const auto nypts = yold.size();
+
+  const auto ncalc = numberCalculated(nypts, stepSize);
+  std::vector<double> xc(ncalc), yc(ncalc);
+  for (size_t step = 0, i = 0; step < nypts; step += stepSize, ++i) {
+    xc[i] = xold[step];
+    yc[i] = yold[step];
+  }
+  // Ensure we have the last value
+  xc.back() = xold.back();
+  yc.back() = yold.back();
+
+  // Manage gsl memory with unique_ptrs
+  using gsl_spline_uptr = std::unique_ptr<gsl_spline, void (*)(gsl_spline *)>;
+  auto spline = gsl_spline_uptr(gsl_spline_alloc(gsl_interp_cspline, ncalc),
+                                [](gsl_spline *sp) { gsl_spline_free(sp); });
```

Comment:
Good point, I'll remove the lambda.

I don't think it's possible to use `make_unique` here as there's no way for it to know what are the arguments to forward to the type and what could be a custom deleter?